FSW/src/util/memUtil.cpp

```diff
@@ -0,0 +1,250 @@
+/* memUtil.cpp handles NS2 payload memory allocation and assignment 
+ * Usage:
+ *  module functionality
+ *  function definitions
+ *  put function declarations in memUtil.hpp
+ * 
+ * Modules encompassed:
+ *  Data Processing
+ *  Science Memory Handling
+ *
+ * Additional files needed for compilation:
+ *  none
+ */
+
+/* - - - - - - Includes - - - - - - */
+// All libraries are put in memUtil.hpp
+// NS2 headers
+#include "../headers/memUtil.hpp"
+
+/* Module Variable Definitions */
+//TODO: should sampling rate and window length be configurable on flight?
+const int SAMPLING_RATE = 50;   // Hz, desired irradiance sampling rate
+const int WINDOW_LENGTH = 240;  // seconds, length of science data buffer 
+const int MAXFILES = 10;        // maximum number of files in flash storage
+
+const int ADC_CHIP_SELECT = 10; // chip select pin for ADC
+const int SPI_MAX_SPEED = 2000000; // clock speed of 2MHz for SPI
+
+// TODO: Update this with size of actual timestamp once it is known
+const int TIMESTAMP_SIZE = 1;   // array indices needed to store timestamp
+
+// contants for converting ADC bins to voltage
+const float ADC_BINS = 65536;       // bins, number of bins in ADC (2^16)
+const float ADC_MAX_VOLTAGE = 3.3;  // Volts, upper end of ADC voltage range
+const float ADC_MIN_VOLTAGE = 0.0;  // Volts, lower end of ADC voltage range
+
+// set number of measurements to store in science data buffer
+const int BUFFERSIZE = SAMPLING_RATE * WINDOW_LENGTH; // indices
+const int FILESIZE = BUFFERSIZE + TIMESTAMP_SIZE;
+
+// declare static variables so that they do not go away when we leave this module
+// whole purpose is to make this into a black-box module of sorts
+static int bufIdx = 0;               // index of next dataBuffer element to overwrite
+static float dataBuffer[BUFFERSIZE]; // create array to hold data buffer elements
+
+
+/* - - - - - - Module Driver Functions - - - - - - */
+
+/* - - - - - - dataProcessing - - - - - - *
+ * Usage:
+ *  Reads incoming data from ADC, measures against baseline,
+ *  sends to memory.
+ * 
+ * Inputs:
+ *  none
+ *  
+ * Outputs:
+ *  data from the ADC (voltage measurement)
+ */
+float dataProcessing()
+{
+    float voltage;
+
+    // begin an SPI connection
+    // TODO: we should probably begin the SPI connection in main to avoid duplicates
+    SPI.begin();  
+    pinMode(ADC_CHIP_SELECT, OUTPUT); // set ADC chip select pin to output
```

Comment:
All the pin modes should be set in main.cpp as well, since I don't think there is ever a reason to change them.

Reply:
You would be correct. The CS pins are only ever pulled low (output). The MOSI lines are only ever master out (output) and the MISO lines are only ever master in (input)